core/mapi/cache/namedPropCache.cpp

```diff
@@ -31,88 +34,182 @@ namespace cache
 					mapi::setTag(lpTag, i) = PROP_TAG(PT_NULL, tag);
 				}
 
-				auto names = GetNamesFromIDs(lpMAPIProp, &lpTag, NULL);
+				const HRESULT hRes = WC_H(GetNamesFromIDs(lpMAPIProp, lpTag, NULL, names));
 				MAPIFreeBuffer(lpTag);
-				return names;
+				return hRes;
 			}
 
-			return {};
+			return MAPI_E_NOT_ENOUGH_MEMORY;
 		}
 
-		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>> GetAllNamesFromIDs(_In_ LPMAPIPROP lpMAPIProp)
+		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>> GetAllNamesFromIDs(_In_ LPMAPIFOLDER lpMAPIFolder)
 		{
 			auto names = std::vector<std::shared_ptr<namedPropCacheEntry>>{};
 
 			// We didn't get any names - try manual
-			const auto ulLowerBound = __LOWERBOUND;
-			const auto ulUpperBound = FindHighestNamedProp(lpMAPIProp);
-			const ULONG batchSize = registry::namedPropBatchSize;
+			constexpr auto ulLowerBound = __LOWERBOUND;
+			const auto ulUpperBound = FindHighestNamedProp(lpMAPIFolder);
+			ULONG batchSize = registry::namedPropBatchSize;
 
 			output::DebugPrint(
 				output::dbgLevel::NamedProp,
 				L"GetAllNamesFromIDs: Walking through all IDs from 0x%X to 0x%X, looking for mappings to names\n",
 				ulLowerBound,
 				ulUpperBound);
-			for (auto iTag = ulLowerBound; iTag <= ulUpperBound && iTag < __UPPERBOUND; iTag += batchSize - 1)
+
+			auto iTag = ulLowerBound;
+			while (iTag <= ulUpperBound && iTag < __UPPERBOUND)
 			{
-				const auto range = GetRange(lpMAPIProp, iTag, min(iTag + batchSize - 1, ulUpperBound));
+				auto end = min(iTag + batchSize - 1, ulUpperBound);
+				std::vector<std::shared_ptr<namedPropCacheEntry>> range;
+				range.reserve(batchSize);
+
+				// Trying to get a range of these props can fail with MAPI_E_CALL_FAILED if it's too big for the buffer
+				// In this scenario, reopen the object, lower the batch size, and try again
+				HRESULT hRes = WC_H(GetRange(lpMAPIFolder, iTag, end, range));
+				if (hRes == MAPI_E_CALL_FAILED)
+				{
+					LPMAPIFOLDER newFolder = nullptr;
+					ULONG objt = 0;
+					hRes = WC_H(lpMAPIFolder->OpenEntry(
+						0, nullptr, &IID_IMAPIFolder, 0, &objt, reinterpret_cast<LPUNKNOWN*>(&newFolder)));
+
+					// If we can't re-open the root folder, something is bad - just break out
+					if (FAILED(hRes))
+					{
+						break;
+					}
+
+					// Swap the folder, lower the batch size, and try again
+					lpMAPIFolder->Release();
+					lpMAPIFolder = newFolder;
+					batchSize /= 2;
```

Comment:
Do you think this warrants a less aggressive backoff? Like 75% instead of 50%, etc?

Reply:
50% seems fine, but I do worry what if somehow batchsize drops to 0 cause we keep failing - we should bail out

Reply:
so it's safe to lookup rootfolder on the folder when it failed get getnames rop? That doesn't autofail like the getnames rop does?

Reply:
@stephenegriffin it certainly doesn't seem to fail like a second call to GetNamesFromIds does. I'll check for the page size getting too low and then bail out.